feat: Transmissibility output

MelReyCG opened the pull request:
This PR is after the #3247.

------------

This PR aims to add a new component: the `StencilOutput`, that allow to add a `CellToCellOutput` task node in the xml to output `ConnectionData` (global id pairs, transmissibilities).

**For now**, by using the `writeCSV="1"` attribute, the `CellToCellOutput` will output this data in a CSV. Here is an exemple:

- XML excerpt
```xml
  <Solvers gravityVector="{0.0, 0.0, 0.0}">
    <SinglePhaseFVM
      name="singlePhaseFlowSolver"
      […]
    </SinglePhaseFVM>
  </Solvers>

  <Tasks>
    <CellToCellOutput
      name="stencilOutput"
      flowSolverName="singlePhaseFlowSolver"
      writeCSV="1"
      logLevel="1" />
  </Tasks>

  <Events maxTime="10.0">
    <SoloEvent
      name="stencilOutputEvent"
      targetCycle="1"
      target="/Tasks/stencilOutput" />
  </Events>
```

- First 10 lines of the generated HDF5
```csv
time [s] , A global id , B global id , AB transmissibility , BA transmissibility
       4 ,           0 ,        2715 , 7.55463e-18                          , -7.55463e-18                        
       4 ,           0 ,       47546 , 8.06167e-18                          , -8.06167e-18                        
       4 ,           0 ,      134459 , 6.63826e-18                          , -6.63826e-18                        
       4 ,           0 ,      165202 , 7.78139e-18                          , -7.78139e-18                        
       4 ,           1 ,        4642 , 1.83325e-17                          , -1.83325e-17                        
       4 ,           1 ,       53907 , 1.76737e-17                          , -1.76737e-17                        
       4 ,           1 ,      184341 , 1.81705e-17                          , -1.81705e-17                        
       4 ,           2 ,        5459 , 1.7414e-17                           , -1.7414e-17                         
       4 ,           2 ,       63907 , 1.86145e-17                          , -1.86145e-17                        
       4 ,           2 ,      177257 , 1.86145e-17                          , -1.86145e-17                        
```
_(aligned here for readability)_

MelReyCG committed:
added transmissibility unit

MelReyCG committed:
Added Cell-To-Cell output

MelReyCG committed:
generated docs

MelReyCG committed:
uncrustify

MelReyCG committed:
Reorganized functions / methods

MelReyCG committed:
removed old method version

MelReyCG committed:
Add the logLevel viewkey & description (cherrypick from feature/rey/sourceflux-stats)

MelReyCG committed:
added wrapper descriptions

MelReyCG committed:
logLevel refactor generated documentation

paveltomin commented on src/coreComponents/physicsSolvers/fluidFlow/StencilOutput.cpp:
```diff
@@ -0,0 +1,233 @@
+      {
+        // gather
+        auto const stencilWrapper = stencil.createKernelWrapper();
+        array1d< KernelConnectionData > const kernelData = gatherTimeStepData( mesh, stencilWrapper );
```
the naming is a bit confusing - why it is time step data?

paveltomin commented on src/coreComponents/physicsSolvers/fluidFlow/StencilOutput.cpp:
```diff
@@ -0,0 +1,233 @@
+    outputFile << time << ","
+               << conn.globalId[0] << ","
+               << conn.globalId[1] << ","
+               << conn.transmissibility[0] << ","
```
not sure it makes much sense to print the same transmissibility value twice but it's ok, not a big deal

karimifard replied:
Maybe it will be more useful to output two half-transmissibilities.

paveltomin commented on src/coreComponents/physicsSolvers/fluidFlow/StencilOutput.cpp:
```diff
@@ -0,0 +1,233 @@
+      FiniteVolumeManager const & fvManager = numericalMethodManager.getFiniteVolumeManager();
+      FluxApproximationBase const & fluxApprox = fvManager.getFluxApproximation( m_solver->getDiscretizationName() );
+
+      fluxApprox.forStencils< CellElementStencilTPFA >( mesh, [&]( auto const & stencil )
```
could you please check if all that will work for https://github.com/GEOS-DEV/GEOS/blob/develop/inputFiles/compositionalMultiphaseFlow/co2_hybrid_1d.xml ?
not sure hybrid discretization has the same things as TPFA

dkachuma replied:
I suppose the key here is that `fvManager.getFluxApproximation( m_solver->getDiscretizationName() );` might return null if the solver doesn't have a flux approximation. So I would add a validator (`postProcessInput()`) to the `solverName` field of this to ensure that the solver provided has the appropriate flux approximation.

dkachuma commented on src/coreComponents/physicsSolvers/fluidFlow/StencilOutput.cpp:
```diff
@@ -0,0 +1,233 @@
+  }
+
+private:
+  StencilOutputKernel()
```
` = delete`?

dkachuma commented on src/coreComponents/physicsSolvers/fluidFlow/StencilOutput.hpp:
```diff
@@ -0,0 +1,169 @@
+  /**
+   * @brief the list of the current registered output target filepaths.
+   */
+  std::set< string > outputFiles;
```
I was a bit confused by this one. Maybe add the `m_` prefix to make it easier to work out that it's a member variable.

dkachuma commented on src/coreComponents/physicsSolvers/fluidFlow/StencilOutput.cpp:
```diff
@@ -0,0 +1,233 @@
+                             real64 const GEOS_UNUSED_PARAM( eventProgress ),
+                             DomainPartition & domain )
+{
+  if( m_writeCSV > 0 && MpiWrapper::commRank() == 0 )
```
Does it mean the root process has all the connections?

MelReyCG replied:
Despite my proposal is not totally finished, it now takes into account all connections within all ranks.

dkachuma commented on src/coreComponents/physicsSolvers/fluidFlow/StencilOutput.cpp:
```diff
@@ -0,0 +1,233 @@
+               << conn.globalId[0] << ","
+               << conn.globalId[1] << ","
+               << conn.transmissibility[0] << ","
+               << conn.transmissibility[1] << std::endl;
```
I would avoid the flush that is implied by `std::endl` and use `\n` - it might speed things up a fraction.

CusiniM reviewed (commented):
It seems that we are trying to only write a file for rank 0 but I don't see when you gather all stencil data there? This would only work for small cases.... 
In any case, why not using the timehistory and writing in a parallel format instead? @wrtobin opinions? 

Besides debugging I don't really see why someone would want to write out transmissibilities.

MelReyCG committed:
Refactor the StencilOutput so the PackCollection can read it.
Point is to allow to do an HDF output, and remove the CSV one.

MelReyCG committed:
Changing the Problem sub-groups init order so the outputs are at the end.
Point is to solve a bug where the StencilDataCollection buffer size are not known for the HDF chunk sizes.

MelReyCG committed:
Joining an uml diagram of the stencil architecture.

MelReyCG committed:
Merge remote-tracking branch 'origin/develop' into feature/rey/transmissibilityOutput

MelReyCG committed:
loglevel viewkeystruct adjustments

MelReyCG committed:
Solved a bug where datasets were duplicated in the hdf5

MelReyCG committed:
Merge remote-tracking branch 'origin/develop' into feature/rey/transmissibilityOutput

MelReyCG committed:
completing merge

MelReyCG committed:
completing merge

MelReyCG committed:
adressing a few review comments

MelReyCG commented:
@CusiniM 
> It seems that we are trying to only write a file for rank 0 but I don't see when you gather all stencil data there? This would only work for small cases.... In any case, why not using the timehistory and writing in a parallel format instead? @wrtobin opinions?
> 
> Besides debugging I don't really see why someone would want to write out transmissibilities.

This proposal is not completely finished, but it now takes into account all connections within all ranks. The `CellToCellDataCollection` now can be used for bigger cases.

To explain why we want to output this data, the goal is to compare the transmissibility data from other simulator(s), in order to understand the differences between the simulations.

MelReyCG committed:
adding a side note

MelReyCG committed:
continuing wip

MelReyCG committed:
CellToCellDataCollection don't support multiple mesh discretization for now

MelReyCG committed:
Merge commit 'c5df35bbf68888cbea60702c1b707a184648e7b7' into feature/rey/transmissibilityOutput

MelReyCG committed:
Revert unrelated modifications

MelReyCG committed:
Adding unit test (has an issue for now)

MelReyCG committed:
Merge branch 'feature/rey/transmissibilityOutput' of https://github.com/GEOS-DEV/GEOS into feature/rey/transmissibilityOutput

MelReyCG committed:
fixing unit test expectations

MelReyCG committed:
generated files

MelReyCG committed:
unit test documentation

MelReyCG committed:
unit test slight refactor to improve code clarity

MelReyCG committed:
Adding a test where cell size is different for each axis

MelReyCG committed:
optimising test + clearing debug logging

rrsettgast commented:
@MelReyCG Can you resolve conflicts?

CusiniM commented on src/coreComponents/finiteVolume/docs/StencilModel.plantuml:
Can you remove these? Coz they are also written in French.

MelReyCG replied:
Are we keeping this in the repo if I translated them?

CusiniM commented on src/coreComponents/physicsSolvers/fluidFlow/StencilDataCollection.cpp:
```diff
@@ -0,0 +1,273 @@
+      FiniteVolumeManager const & fvManager = numericalMethodManager.getFiniteVolumeManager();
+      FluxApproximationBase const & fluxApprox = fvManager.getFluxApproximation( m_solver->getDiscretizationName() );
+
+      fluxApprox.forStencils< CellElementStencilTPFA >( mesh, [&]( auto const & stencil )
```
Maybe supporting the output for all the other stencils could be useful too. I m thinking of debugging of fractured problems.

MelReyCG replied:
Of course, I don't currently have any cases with fractures that need this output, so I may add more supported stencils later, but this component is opened to addition!

rrsettgast commented:
@MelReyCG Can you merge develop into this?

MelReyCG committed:
Changed the way to get mesh & discretization objects

MelReyCG committed:
Translated forgotten french notes

MelReyCG committed:
bugfix

MelReyCG committed:
unit test adaptations

MelReyCG committed:
Merge commit 'a7ddc53638bc37ac1927904e0c1d4e4dad754cda' into feature/rey/transmissibilityOutput

MelReyCG committed:
postProcessInput -> postInputInitialization

rrsettgast committed:
Merge branch 'develop' into feature/rey/transmissibilityOutput

MelReyCG committed:
Merge branch 'develop' into feature/rey/transmissibilityOutput

MelReyCG committed:
update submodule

MelReyCG committed:
🐛 exposing members for CUDA kernel compilation

MelReyCG committed:
gpu bugfix: removing accessing by function

MelReyCG committed:
update submodules

MelReyCG committed:
Merge branch 'develop' into feature/rey/transmissibilityOutput

MelReyCG committed:
Merge branch 'develop' into fileIO-refactor-start

MelReyCG committed:
removing host calls from host-device

MelReyCG committed:
📦 generated files update

MelReyCG committed:
removing full context for simple loggings

MelReyCG committed:
Joining an exemple to illustrate how to use the component.

MelReyCG committed:
todos

MelReyCG committed:
adding some info to error message

MelReyCG committed:
code style + more info in kernel error msg

MelReyCG committed:
temporary debug info

MelReyCG committed:
adding more temporary debug info

MelReyCG committed:
adding more debug info

MelReyCG committed:
adding more debug info

MelReyCG committed:
more and more debug infos

MelReyCG committed:
debug info uncrustify

MelReyCG committed:
silly fix test

MelReyCG committed:
freeOnDevice fix test

MelReyCG committed:
Merge remote-tracking branch 'origin/develop' into feature/rey/transmissibilityOutput

MelReyCG committed:
DataTypes.hpp now only manage static types
- no runtime types nor Mpi constants
- DataTypes.cpp is not useful anymore.

MelReyCG committed:
generated files

MelReyCG committed:
Merge remote-tracking branch 'origin/develop' into fileIO-refactor-basicIO&format-under-common

MelReyCG committed:
🐛 physicsSolvers refactor compilation fix

MelReyCG committed:
sharper todos

MelReyCG committed:
♻️🔇 removing debug outputs + prefering move(host) instead of freeOnDevice()

MelReyCG committed:
more todo

MelReyCG committed:
removing common -> fileIO

MelReyCG committed:
last common->fileIO ref

MelReyCG committed:
removing common->codingUtilities

MelReyCG committed:
logger has now a dedicated folder for upcoming files (future PRs)

MelReyCG committed:
🐛 compil fix (dynamicCast)

MelReyCG committed:
📝 missed docs (where not previously catched by the CI)

MelReyCG committed:
🐛 compil fix fix (dynamicCast)

MelReyCG committed:
📝 docs fix

MelReyCG committed:
Merge branch 'develop' into feature/rey/remove-circular-fileIO&common-dependancies

MelReyCG committed:
Merge branch 'develop' into feature/rey/transmissibilityOutput

MelReyCG committed:
Merge branch 'develop' into feature/rey/transmissibilityOutput

MelReyCG committed:
Merge remote-tracking branch 'origin/develop' into feature/rey/remove-circular-fileIO&common-dependancies

MelReyCG committed:
Merge remote-tracking branch 'origin/develop' into feature/rey/remove-circular-fileIO&common-dependancies

MelReyCG commented:
Just leaving a note here, no CI step is legitimately failing here. The CUDA builds were working but crash now because of a CI bug since I updated my branch.

MelReyCG committed:
Merge branch 'develop' into feature/rey/transmissibilityOutput

MelReyCG committed:
Merge remote-tracking branch 'origin/feature/rey/remove-circular-fileIO&common-dependancies' into feature/rey/transmissibilityOutput

MelReyCG committed:
Merge branch 'feature/rey/transmissibilityOutput' of https://github.com/GEOS-DEV/GEOS into feature/rey/transmissibilityOutput

CusiniM commented:
This PR needs one more review.

rrsettgast committed:
Merge branch 'develop' into feature/rey/transmissibilityOutput

rrsettgast commented on src/coreComponents/dataRepository/Group.hpp:
```diff
@@ -338,8 +338,12 @@ class Group
                    std::domain_error );
-
-    return dynamicCast< T & >( *child );
+    T * const castedChild = dynamicCast< T * >( child );
+    GEOS_THROW_IF( castedChild == nullptr,
+                   GEOS_FMT( "{} was expected to be a '{}'.",
+                             child->getDataContext(), LvArray::system::demangleType< T >() ),
+                   BadTypeError );
+    return *castedChild;
```
There is a `dynamic_cast` function in `RTTypes.hpp` for this.
```c++
template< typename NEW_TYPE, typename EXISTING_TYPE >
NEW_TYPE dynamicCast( EXISTING_TYPE & val )
{
  static_assert( std::is_reference< NEW_TYPE >::value, "NEW_TYPE must be a reference." );

  using POINTER_TO_NEW_TYPE = std::remove_reference_t< NEW_TYPE > *;
  POINTER_TO_NEW_TYPE ptr = dynamicCast< POINTER_TO_NEW_TYPE >( &val );
  GEOS_ERROR_IF( ptr == nullptr, "Cast from " << LvArray::system::demangleType( val ) << " to " <<
                 LvArray::system::demangleType< NEW_TYPE >() << " failed." );

  return *ptr;
}
```
